lib/ukfs: Unikraft Filesystem Interface

andreittr opened the pull request:
### Description of changes

This PR introduces `ukfs`, the core filesystem API for `ukfile`s, enabled on-demand by selecting the `ukfs` library for build.
This adds a "filesystem ops" table reference to every file, with "true filesystem" files having it populated by their driver, and pseudo-files setting it to `NULL`.
Existing file implementations that initialize their `struct uk_file` with designated initializers (all from what I can tell) are automatically backwards-compatible.

This PR provides no drivers or callers into the `ukfs` API; rather it serves as a point of discussion of the interface itself and its logical consistency.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

andreittr committed:
lib/ukfs: Introduce ukfs library

This change introduces ukfs, the Unikraft filesystem interface lib.

Checkpatch-Ignore: ENOSYS
Checkpatch-Ignore: AVOID_EXTERNS
Checkpatch-Ignore: PREFER_KERNEL_TYPES
Checkpatch-Ignore: REPEATED_WORD
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

mogasergiu commented on lib/ukfs/Makefile.uk:
```diff
@@ -0,0 +1,9 @@
+LIBUKFS_SRCS-y += $(LIBUKFS_BASE)/common-ops.c
+LIBUKFS_SRCS-y += $(LIBUKFS_BASE)/drivers.c
+
+LIBUKFS_SRCS-y += $(LIBUKFS_BASE)/driver-list.ld
```
I don't see any other source file with `-` in the name in our codebase, feels inconsistent :thinking:
Also maybe this should have used the `.lds.S` suffix instead?

andreittr replied:
> I don't see any other source file with - in the name in our codebase, feels inconsistent

[Plenty](https://github.com/unikraft/unikraft/blob/staging/lib/ukfile/file-nops.c) of [examples](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fd/include/uk/posix-fd.h) across [much](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fdio/fd-shim.c) of the [new](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fdio/fdio-impl.h) file/vfs [stack](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fdio/include/uk/posix-fdio.h). This is following the same pattern.

> Also maybe this should have used the .lds.S suffix instead?

I'm not sure; [other](https://github.com/unikraft/unikraft/blob/staging/lib/ukalloc/libstats.ld) libs [also](https://github.com/unikraft/unikraft/blob/staging/lib/posix-process/clonetab.ld) use `.ld` for linker scripts, though `.lds.S` is also used in about as many places :shrug:

mogasergiu replied:
> [Plenty](https://github.com/unikraft/unikraft/blob/staging/lib/ukfile/file-nops.c) of [examples](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fd/include/uk/posix-fd.h) across [much](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fdio/fd-shim.c) of the [new](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fdio/fdio-impl.h) file/vfs [stack](https://github.com/unikraft/unikraft/blob/staging/lib/posix-fdio/include/uk/posix-fdio.h). This is following the same pattern.

Introduced by you.. :sweat_smile: , ok sure, fine.

Ok so this was inconsistent from the very start:
ld:
```
./lib/ukdebug/trace.ld           
./lib/ukdebug/gdbtargetxml.ld    
./lib/vfscore/extra.ld           
./lib/uktest/tests.ld            
./lib/posix-socket/driver_list.ld
./lib/ukalloc/libstats.ld        
./lib/uknofault/excpttab.ld      
./lib/posix-process/clonetab.ld  
./lib/uksched/extra.ld           
./lib/uklibid/infosec.ld         
```
lds.S:
```
./plat/kvm/x86/link64.lds.S               
./plat/kvm/arm/link64.lds.S               
./plat/xen/x86/link64.lds.S               
./plat/xen/arm/link32.lds.S               
./plat/xen/arm/link64.lds.S               
./plat/common/bootinfo.lds.S              
./lib/uklibparam/libparam.lds.S           
./lib/syscall_shim/syscall_exittab.lds.S  
./lib/syscall_shim/syscall_entertab.lds.S 
./lib/ukboot/earlytab.lds.S               
./lib/ukreloc/reloc.lds.S                 
./lib/ukgcov/libukgcov.lds.S              
```
Ok, fine... can't help but feel this is very inconsistent.

mogasergiu commented on lib/ukfs/drivers.c:
```diff
@@ -0,0 +1,38 @@
+uk_fs_vopen_func uk_fs_driver(const char *fstype)
+{
+	const struct uk_fs_drv *drv = uk_fs_driver_list_start;
+	size_t len = UKFS_DRIVER_COUNT;
```
`__sz`? Ditto for other occurences of `size_t`, especially since it's supposed to be Unikraft internal.

andreittr replied:
I disagree, `ukfs` builds on top of `ukfile`, which by design always pulls in a libc and uses libc types; and if we're in a libc-enabled env anyway, I don't like mixing libc and `__` types in the same source file.
This dependency on libc types goes beyond primitives and into specific structs (e.g., `struct statfs` used directly by `ukfs`, with others used by `ukfile`) so it's not something easy to remove.

mogasergiu commented on lib/ukfs/include/uk/fs/driver.h:
```diff
@@ -0,0 +1,46 @@
+ */
+#define UK_FS_DRIVER_REGISTER(fsname, drv_vopen)		\
+	__used __align(8)					\
+		__section("." STRINGIFY(uk_fs_driver_##fsname))	\
```
Interesting, so each driver declares its own section that will be part of `uk_fs_driver_list`. Is there a reason for this? :thinking:

andreittr replied:
The reason is so that linker magic will sort these structs by name in the array, allowing us to do binary search + strcmp.
The magic itself happens by sorting the drivers' sections at link time.

mogasergiu reviewed (commented):
Very nice addition, can't wait to see first registered driver. In the meantime I had a brief read and only a few simple questions, since there is not much runtime code to look at **yet**.

andreittr commented:
Pushed revision of special file API.

andreittr commented:
Pushed rebase to `RELEASE-0.19.0`.

andreittr commented:
Pushed minor change to `create` API.

mogasergiu commented on lib/ukfs/include/uk/fs/driver.h:
```diff
@@ -0,0 +1,46 @@
+#define UK_FS_DRIVER_REGISTER(fsname, drv_vopen)		\
+	__used __align(8)					\
+		__section("." STRINGIFY(uk_fs_driver_##fsname))	\
+	static const struct uk_fs_drv uk_fs_driver_##fsname = {	\
```
This looks like a constant registration that stays constant. Any reason the driver list section is inserted after `.data` instead of `.rodata`?

andreittr replied:
:eyes: good catch, it does make more sense to be after `.rodata`.
I must've uncritically copied the linker script of [posix-socket](https://github.com/unikraft/unikraft/blob/staging/lib/posix-socket/driver_list.ld#L10) and that indeed has non-const driver entries (_why_ that is eludes me though).

andreittr commented:
Pushed update addressing comments.

mogasergiu commented on lib/ukfs/include/uk/fs/driver.h:
```diff
@@ -0,0 +1,46 @@
+ *  != NULL: vopen function of desired filesystem
+ *  == NULL: Filesystem name `fstype` not recognized
+ */
+uk_fs_vopen_func uk_fs_driver(const char *fstype);
```
This is technically not retrieving the driver itself but its `vopen` function :thinking: . Why not `uk_fs_vopen_by_name` or something like that.

andreittr replied:
In ukfs, at a very fundamental level, a driver _is_ its vopen function, and there is no distinction between the two -- we have no centralized ops table or anything, all driver functionality is encapsulated in the files it produces. `vopen` is the seed from which you get your first file. This is something I should probably insist on when doing the high-level doc pass over ukfile/ukfs :smile: .

Furthermore, at this very high abstraction layer the naming should best reflect the intentions of the caller -- to find the ukfs driver for `fstype` -- and be less about the details of _how_ a driver is represented.

If we ever want ukfs drivers to expose more metadata about themselves (and thus break the very strict driver-vopen equivalence mentioned above), we'll then define a
```c
struct uk_fs_driver_info {
    uk_fs_vopen_func vopen;
    /* misc metadata */
};
```
have drivers register using that, and change to
```c
const struct uk_fs_driver_info *uk_fs_driver(const char *fstype);
```
thus keeping an appropriate abstraction.
(We don't do this struct thing now to prevent an unnecessary extra indirection)

mogasergiu commented on lib/ukfs/include/uk/fs/driver.h:
```diff
@@ -0,0 +1,46 @@
+		__section("." STRINGIFY(uk_fs_driver_##fsname))	\
+	static const struct uk_fs_drv uk_fs_driver_##fsname = {	\
+		.fstype = STRINGIFY(fsname),			\
+		.vopen = drv_vopen				\
```
Would it make sense to surround `drv_vopen` with paranthesis?

andreittr replied:
:thinking: yeah, looks like parantheses won't break anything here, must've been an oversight :+1:

mogasergiu commented on lib/ukfs/include/uk/fs.h:
```diff
@@ -0,0 +1,591 @@
+/* Non-directories must not implement `listdir` */
+#define UKFS_FILE_ISDIR(f) (UKFS_FILE_ISFS(f) && (f)->fsops->listdir)
+/* Only symlinks may implement `readlink` */
+#define UKFS_FILE_ISSYM(f) (UKFS_FILE_ISFS(f) && (f)->fsops->readlink)
```
I see multi-eval macro again where someone could pass a function call as an argument. Would it be worth making it a static inline instead?

andreittr replied:
Yeah, I guess static inlines would work just as well here and be more typesafe :+1: .

andreittr commented:
Pushed addressing comments.

andreittr commented:
Pushed minor change reordering the fsops struct to have the most commonly used operations at the start. API remains unchanged.

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented:
Pushed update adding header `uk/fs/prio.h` to synchronize init/term prio among fs drivers & consumers thereof. Re-requesting review for just the additional header.

andreittr commented:
Pushed rebase to current staging.

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented:
Pushed very minor correction to the public API:
`vopen` and its function type is now declared in `uk/fs/driver.h`, as it it relevant only when talking about fs drivers:
- A driver implementing its vopen function
- A ukfs caller that wishes to look up a driver by name with `uk_fs_driver`

A caller only interested in filesystem operations (including the main `uk/fs.h` header) need not know or care about `vopen`.

andreittr commented:
Pushed update with 2 changes:
- added `exportsyms.uk` file to explicitly express developer intentions
- squashed the previous 2 commits into one, as the intermediate state would lead to a broken build with ukfs enabled

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
